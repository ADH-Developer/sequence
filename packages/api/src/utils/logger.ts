import winston from "winston";

const logger = winston.createLogger({
  level: "info",
  format: winston.format.timestamp(),
  transports: [new winston.transports.Console()],
});

export default logger;
