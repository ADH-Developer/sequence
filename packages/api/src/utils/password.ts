import crypto from "crypto";
import { promisify } from "util";

const scrypt = promisify(crypto.scrypt);

export async function hash(password: string) {
  const salt = crypto.randomBytes(8).toString("hex");

  const derivedKey = (await scrypt(password, salt, 64)) as Buffer;
  return salt + ":" + derivedKey.toString("hex");
}

export async function verify(password: string, hash: string) {
  const [salt, key] = hash.split(":");
  const keyBuffer = Buffer.from(key, "hex");
  const derivedKey = (await scrypt(password, salt, 64)) as Buffer;
  return crypto.timingSafeEqual(keyBuffer, derivedKey);
}
