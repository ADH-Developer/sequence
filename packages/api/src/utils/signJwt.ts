import jwt from "jsonwebtoken";
import User from "src/models/user.model";
import JwtConfig from "src/config/jwt";

export const signJwt = (user: User): Promise<string> => {
  return new Promise((resolve, reject) => {
    jwt.sign(
      {
        user,
      },
      JwtConfig.jwt.secret,
      JwtConfig.jwt.options,
      (error: Error, encoded: string) => {
        if (error) {
          reject(error);
        }
        resolve(encoded);
      }
    );
  });
};
