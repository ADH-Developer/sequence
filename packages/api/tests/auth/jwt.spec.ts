import { gql } from "apollo-server-express";
import { expect } from "chai";
import { stub } from "sinon";
import {
  createTestClient,
  TestQuery,
  TestSetOptions,
} from "apollo-server-integration-testing";
import App from "src/app";
import User from "src/models/user.model";
import { signJwt } from "src/utils/signJwt";
import userSeed from "tests/seeds/user.seed";

let app: App;
let user: User;
let jwt: string;
let query: TestQuery;
let mutate: TestQuery;
let setOptions: TestSetOptions;

describe("jwt authentication", function () {
  before(async () => {
    app = new App();
    user = await userSeed({
      firstName: "Test",
      lastName: "User",
    });
    jwt = await signJwt(user);
    const result = createTestClient({
      //@ts-ignore
      apolloServer: app.apolloServer,
    });
    query = result.query;
    mutate = result.mutate;
    setOptions = result.setOptions;
  });

  it("should reject no token provided", async () => {
    try {
      await query(gql`
        query TestQuery {
          getUser {
            firstName
            lastName
          }
        }
      `);
    } catch (error) {
      expect(error).toMatchSnapshot();
    }
  });
  it("should reject bad token provided", async () => {
    setOptions({
      request: {
        headers: {
          authorization: "Bearer my_invalid_token",
        },
      },
    });
    try {
      await query(gql`
        query TestQuery {
          getUser {
            firstName
            lastName
          }
        }
      `);
    } catch (error) {
      expect(error).toMatchSnapshot();
    }
  });
  it("should accept a JWT", async () => {
    setOptions({
      request: {
        headers: {
          authorization: `Bearer ${jwt}`,
        },
      },
    });
    try {
      const result = await query(gql`
        query TestQuery {
          getUser {
            firstName
            lastName
          }
        }
      `);
      expect(result).toMatchSnapshot();
    } catch (error) {
      expect(error).to.be.undefined;
    }
  });
});
