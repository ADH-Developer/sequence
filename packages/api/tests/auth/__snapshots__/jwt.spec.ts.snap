// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jwt authentication should accept a JWT 1`] = `
Object {
  "data": Object {
    "getUser": Object {
      "firstName": "Test",
      "lastName": "User",
    },
  },
}
`;

exports[`jwt authentication should reject bad token provided 1`] = `
[HttpQueryError: {"errors":[{"message":"Context creation failed: You are not authorized to perform this request.","errors":null}]}
]
`;

exports[`jwt authentication should reject no token provided 1`] = `
[HttpQueryError: {"errors":[{"message":"Context creation failed: Authentication not supported","errors":null}]}
]
`;
